### PR TITLE
GUI Utilities: Implement PS3 SDAT/EDAT decryption

### PR DESCRIPTION
**Pull request by elad335**
In addition to PPU SELF/SPRX/OVL decryption, the tool "Decrypt PS3 Binaries" will now be able to decrypt SDAT and EDAT files.
Note that EDAT files may require KLIC and RAP file, I've added logging of the needed licence file (NPD content_id) for such cases.

### CODE REVIEW

File: rpcs3/Crypto/unedat.cpp [hunk 1]
```diff
@@ -715,26 +716,31 @@ bool extract_all_data(const fs::file* input, const fs::file* output, const char*
...
+			if (fs::file rap{rap_path}; rap && rap.size() >= sizeof(key))
 			{
-				edat_log.error("A valid RAP file is needed for this EDAT file! (local activation)");
-				return true;
+				key = GetEdatRifKeyFromRapFile(rap);
+			}
+			else if (rifkey)
+			{
+				// Fallback
+				if (rifkey)
```
Comment (Megamouse) [inline rpcs3/Crypto/unedat.cpp:733]: double check

File: rpcs3/Crypto/unedat.cpp [hunk 2]
```diff
@@ -715,26 +716,31 @@ bool extract_all_data(const fs::file* input, const fs::file* output, const char*
...
+				{
+					std::memcpy(&key, rifkey, 0x10);
+				}
+
```
Comment (Megamouse) [inline rpcs3/Crypto/unedat.cpp:737]: empty line

File: rpcs3/Crypto/unedat.cpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Crypto/unedat.cpp:733]: I don't know why it even exists as a param if it's not used anymore

File: rpcs3/Crypto/unedat.cpp [hunk 3]
```diff
@@ -715,26 +716,31 @@ bool extract_all_data(const fs::file* input, const fs::file* output, const char*
...
 		if ((NPD.license & 0x3) == 0x3)           // Type 3: Use supplied devklic.
-			memcpy(&key, devklic, 0x10);
-		else if ((NPD.license & 0x2) == 0x2)      // Type 2: Use key from RAP file (RIF key).
 		{
-			memcpy(&key, rifkey, 0x10);
+			std::memcpy(&key, devklic, 0x10);
+		}
+		else // Type 2: Use key from RAP file (RIF key). (also used for type 1 at the moment)
+		{
+			const std::string rap_path = rpcs3::utils::get_rap_file_path(NPD.content_id);
```
Comment (Megamouse) [inline rpcs3/Crypto/unedat.cpp:724]: I have concerns that this will make everything slower, since it will now look for a rap file everytime.

Comment (elad335) [reply rpcs3/Crypto/unedat.cpp:724]: DecryptEDAT has always read RAP file, and this function is only used by it.

Comment (Megamouse) [reply rpcs3/Crypto/unedat.cpp:724]: no. it was only reading a rap file if provided. now it always searches for the rap file, which it clearly wasn't doing before.

Comment (Megamouse) [reply rpcs3/Crypto/unedat.cpp:724]: for example package installation

Comment (elad335) [reply rpcs3/Crypto/unedat.cpp:724]: EDAT that needs RAP file will always need it, not providing RAP path in such case will just yield a decryption failure.
And looking for RAP file typically ends within a single iteration if RAP file exists, fs::is_file is not expensive.
If fs::is_file is really a concern get_rap_file_path can return fs::file to it as well as path.

Comment (elad335) [reply rpcs3/Crypto/unedat.cpp:724]: > https://github.com/RPCS3/rpcs3/pull/10934#discussion_r720637085
PKG installation currently only supports SDAT decrytion, which doesn't need RAP and won't reach this code.

Comment (elad335) [reply rpcs3/Crypto/unedat.cpp:724]: The way I see it is that it prevents decryption failure bugs, simplifies usage with very little performance panelty.

File: rpcs3/Crypto/unedat.cpp (hunk 2 above)
Comment (elad335) [reply rpcs3/Crypto/unedat.cpp]: Fixed.

File: rpcs3/Crypto/unedat.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Crypto/unedat.cpp]: Removed.

File: rpcs3/Crypto/unedat.cpp (hunk 3 above)
Comment (elad335) [reply rpcs3/Crypto/unedat.cpp:724]: Removed redundancy in verify_c00_unlock_edat because of it.

Comment (Megamouse) [reply rpcs3/Crypto/unedat.cpp:724]: We'll see if there are regressions ;)

**Pull request merged**